File: typescript/rna-transcription/rna-transcription.ts
const dnaNucleotides = ['A', 'C', 'G', 'T'] as const;

type DNA = typeof dnaNucleotides[number];
type RNA = 'A' | 'C' | 'G' | 'U';
type Complement = Record<DNA, RNA>;

const complements: Complement = {
  G: 'C',
  C: 'G',
  T: 'A',
  A: 'U',
};

const isDnaNucleotide = (s: string): s is DNA => dnaNucleotides.includes(s as DNA);

export const dnaToRnaComplement = (dna: string): string => (
  dna.split('')
    .map((n: string): RNA => {
      if (!isDnaNucleotide(n)) {
        throw new Error('Invalid input DNA.');
      }

      return complements[n];
    })
    .join('')
);

File: scala/rna-transcription/src/main/scala/RnaTranscription.scala
object RnaTranscription {
  // private val complements = Map[String, String](
  //   "G" -> "C",
  //   "C" -> "G",
  //   "T" -> "A",
  //   "A" -> "U"
  // );

  def toRna(dna: String): Option[String] = {
    val complement = dna.map {
      case 'G' => 'C'
      case 'C' => 'G'
      case 'T' => 'A'
      case 'A' => 'U'
    }

    Some(complement)
  }

  // def toRna(dna: String): Option[String] = Some(
  //   dna
  //     .toList
  //     .map(c => complements(c.toString))
  //     .mkString
  // )
}
